app/collections/collaborators.jsx

```diff
@@ -0,0 +1,375 @@
+      } else {
+        errorMessage = 'Error adding user.';
+      }
+    }
+
+    return (
+      <div ref={(node) => { this.node = node; }}>
+        {this.state.error && (
+          <p className="form-help error">{errorMessage}</p>
+        )}
+
+        <form style={style}>
+          <div>
+            <UserSearch ref={(component) => { this.userSearch = component; }} />
+          </div>
+
+          <table className="standard-table">
+            <tbody>
+              {POSSIBLE_ROLES.map(role =>
+                <tr key={role}>
+                  <td><input id={ID_PREFIX + role} type="checkbox" name="role" value={role} disabled={role === 'owner'} /></td>
+                  <td><strong><label htmlFor={ID_PREFIX + role}>{ROLES_INFO[role].label}</label></strong></td>
+                  <td>{ROLES_INFO[role].description}</td>
+                </tr>
+              )}
+            </tbody>
+          </table>
+
+          <p>
+            <button type="submit" className="major-button" onClick={this.handleSubmit}>Add user role</button>
+          </p>
+        </form>
+      </div>
+    );
+  }
+}
+
+RoleCreator.defaultProps = {
+  collections: null
+};
+
+RoleCreator.propTypes = {
+  collection: React.PropTypes.shape({
+    id: React.PropTypes.string
+  }),
+  onAdd: React.PropTypes.func
+};
+
+export class RoleRow extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      owner: {},
+      saving: null
+    };
+
+    this.removeRoles = this.removeRoles.bind(this);
+    this.toggleRole = this.toggleRole.bind(this);
+    this.updateRoles = this.updateRoles.bind(this);
+  }
+
+  componentDidMount() {
+    this.props.roleSet.get('owner')
+      .then((owner) => {
+        this.setState({ owner });
+      });
+  }
+
+  removeRoles() {
+    this.updateRoles([], this.props.onRemove);
+  }
+
+  toggleRole(role) {
+    const currentRoles = this.props.roleSet.roles;
+    const removing = currentRoles.includes(role);
+
+    if (removing) {
+      this.confirmDelete(currentRoles, role);
+    } else {
+      currentRoles.push(role);
+      this.updateRoles(currentRoles);
+    }
+  }
+
+  confirmDelete(currentRoles, role) {
+    alert((resolve) => {
+      const handleDelete = () => {
+        const index = currentRoles.indexOf(role);
+        currentRoles.splice(index, 1);
+
+        if (currentRoles.length) {
+          this.updateRoles(currentRoles);
+        } else {
+          this.removeRoles();
+        }
+        resolve();
+      };
+
+      return (
+        <div className="confirm-delete-dialog content-container">
+          <p>Are you sure you want to delete this role?</p>
+          <div>
+            <button className="minor-button" autoFocus={true} onClick={resolve}>No</button>
+            {' '}
+            <button className="major-button" onClick={handleDelete}>Yes</button>
+          </div>
+        </div>
+      );
+    });
+  }
+
+  updateRoles(newRoles = [], callback = () => {}) {
```

Comment:
The callback is being used right before the `setState` in `updateRoles()` to call the `this.props.onRemove` that is passed to it from the `removeRoles()` function.